modyn/storage/src/internal/database/storage_database_connection.cpp

```diff
@@ -79,45 +79,54 @@ bool StorageDatabaseConnection::add_dataset(const std::string& name, const std::
                                             const FilesystemWrapperType& filesystem_wrapper_type,
                                             const FileWrapperType& file_wrapper_type, const std::string& description,
                                             const std::string& version, const std::string& file_wrapper_config,
-                                            const bool ignore_last_timestamp,
-                                            const int64_t file_watcher_interval) const {
+                                            const bool ignore_last_timestamp, const int64_t file_watcher_interval,
+                                            const bool upsert) const {
   soci::session session = get_session();
 
   auto filesystem_wrapper_type_int = static_cast<int64_t>(filesystem_wrapper_type);
   auto file_wrapper_type_int = static_cast<int64_t>(file_wrapper_type);
   std::string boolean_string = ignore_last_timestamp ? "true" : "false";
 
-  if (get_dataset_id(name) != -1) {
-    SPDLOG_ERROR("Dataset {} already exists", name);
+  const int64_t dataset_id = get_dataset_id(name);
+
+  if (dataset_id != -1 && !upsert) {
+    // Dataset already exists
+    SPDLOG_ERROR("Cannot insert dataset {} as it already exists", name);
     return false;
   }
-  switch (drivername_) {
-    case DatabaseDriver::POSTGRESQL:
-      try {
-        session << "INSERT INTO datasets (name, base_path, filesystem_wrapper_type, "
-                   "file_wrapper_type, description, version, file_wrapper_config, "
-                   "ignore_last_timestamp, file_watcher_interval, last_timestamp) "
-                   "VALUES (:name, "
-                   ":base_path, :filesystem_wrapper_type, :file_wrapper_type, "
-                   ":description, :version, :file_wrapper_config, "
-                   ":ignore_last_timestamp, :file_watcher_interval, 0)",
-            soci::use(name), soci::use(base_path), soci::use(filesystem_wrapper_type_int),
+
+  if (dataset_id != -1) {
+    SPDLOG_INFO("Dataset {} already exists, updating...", name);
+
+    const std::string update_query =
+#include "sql/SQLUpdateDataset.sql"
+        ;
+
+    switch (drivername_) {
+      // same logic for both dbms > fallthrough
+      case DatabaseDriver::POSTGRESQL:
+      case DatabaseDriver::SQLITE3:
```

Comment:
Why do we need the switch statement here (and below), if the SQL is the same in both cases anyways?

Reply:
True, I will get rid of it! :)